### PR TITLE
Refs [TASK] [34082] [Init Remote data source ]

### PR DESCRIPTION
**Pull request by Hoang-Kelvin**
## Related Tickets
- [#Task 34082: Init Remote data source]({https://edu-redmine.sun-asterisk.vn/issues/xxxx})
## WHAT
 - Create request to get data
- Create remote data source
- Create Model
## Review Checklist

Category | View Point | Description | Expected Reviewer Answer | Self review | Reviewer2 (name)
--- | --- | --- | --- | --- | ---
Conventions | Does the code follow Sun* coding style and coding conventions? | https://github.com/framgia/coding-standards/tree/master/eng/android | YES |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
Redmine | Does the ticket follow Sun* Redmine working process?  | https://github.com/framgia/Training-Guideline/blob/master/WorkingProcess/redmine/redmine.md| YES |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
Documentation | Is there any incomplete code? If so, should it be removed or flagged with a suitable marker like ‘TODO’? |  | YES |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
## Notes (Optional)
*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(List gem, library third party add new)*

*(Other notes)*

### CODE REVIEW

File: app/src/main/java/com/project/gamedb/data/model/Games.kt [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        jsonObject.getJSONArray(GAME_GENRES).map(::Genres),
+        if (jsonObject.getString(GAME_META_CRITIC) != "null")
+        jsonObject.getLong(GAME_META_CRITIC) else -1,
+        jsonObject.getLong(GAME_PLAY_TIME),
+        jsonObject.getLong(GAME_SUGGESTION_COUNT),
+        jsonObject.getString(GAME_UPDATED),
+        if (jsonObject.getString(GAME_ESRB_RATING) != "null") jsonObject.getJSONObject(
+            GAME_ESRB_RATING
+        ).let(::ESRBRating) else null,
+        jsonObject.getJSONArray(GAME_PLATFORM).map(::PlatformByGame)
```
Comment (thonc-2135) [inline app/src/main/java/com/project/gamedb/data/model/Games.kt]: xem lại convention vs if else nhé a

File: app/src/main/java/com/project/gamedb/data/source/remote/api/ApiConstants.kt [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package com.project.gamedb.data.source.remote.api
+
+object ApiConstants {
+    const val SCHEME_HTTPS = "https"
+    const val AUTHORITY_API_RAWG = "api.rawg.io"
+    const val GAMES ="games"
+    const val API ="api"
+    const val API_KEY = "6617d6d0760f413d9bd17f35b63d3b71"
```
Comment (thonc-2135) [inline app/src/main/java/com/project/gamedb/data/source/remote/api/ApiConstants.kt]: giấu key đi a , tham khảo nhé:  https://viblo.asia/p/cac-thu-thuat-de-giau-secret-keypassword-khoi-source-control-va-ung-dung-voi-ci-trong-lap-trinh-android-PwRGgmOnkEd

File: app/src/main/java/com/project/gamedb/ultis/JSONExtentions.kt [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+import org.json.JSONArray
+import org.json.JSONObject
+
+inline fun <R> JSONArray.map(transform : (JSONObject) -> R?) : List<R> =
+    mutableListOf<R>().also { list ->
+        repeat(length()){
+            (get(it) as? JSONObject)?.let(transform)?.let(list::add)
+        }
+    }
+fun toStringArray(array:JSONArray):List<String> {
```
Comment (thonc-2135) [inline app/src/main/java/com/project/gamedb/ultis/JSONExtentions.kt]: hàm này là extention function hả a , hình như còn chưa đúng format đó

File: app/src/main/java/com/project/gamedb/data/source/remote/OnDataLoadedCallback.kt [hunk 4]
```diff
@@ -0,0 +1,8 @@
+package com.project.gamedb.data.source.remote
+
+import java.lang.Exception
+
+interface OnDataLoadedCallback<T> {
+    fun onSuccess(data: T)
+    fun onFailure(exception: Exception? = Exception("Unexpected Error"))
```
Comment (thonc-2135) [inline app/src/main/java/com/project/gamedb/data/source/remote/OnDataLoadedCallback.kt]: đây có gọi là hard code k a ?

File: app/src/main/java/com/project/gamedb/ultis/JSONExtentions.kt (hunk 3 above)
Comment (datdt-1935) [inline app/src/main/java/com/project/gamedb/ultis/JSONExtentions.kt]: để thành extension luôn nhé

File: app/src/main/java/com/project/gamedb/ultis/Repositories.kt [hunk 5]
```diff
@@ -0,0 +1,11 @@
+package com.project.gamedb.ultis
+
+import com.project.gamedb.data.source.remote.RemoteDataRepository
+import com.project.gamedb.data.source.remote.RemoteDataSource
+
+object Repositories {
+    fun getRemoteRepository() : RemoteDataRepository{
+        val remote = RemoteDataSource()
```
Comment (datdt-1935) [inline app/src/main/java/com/project/gamedb/ultis/Repositories.kt]: không khởi tạo ở đây mà truyền vào từ constructor

File: app/src/main/java/com/project/gamedb/data/source/GameDataSource.kt [hunk 6]
```diff
@@ -0,0 +1,6 @@
+package com.project.gamedb.data.source
+
+interface GameDataSource {
+    interface Remote{
```
Comment (thonc-2135) [inline app/src/main/java/com/project/gamedb/data/source/GameDataSource.kt]:
```suggestion
    interface Remote {
```

**Pull request merged**